torch/csrc/distributed/spmd/event_handler_impl.cpp

```diff
@@ -0,0 +1,209 @@
+#include <torch/csrc/distributed/spmd/event_handler_impl.h>
+
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/autograd/functions/accumulate_grad.h>
+#include <torch/csrc/autograd/utils/lambda_post_hook.h>
+
+
+namespace torch {
+namespace distributed {
+namespace spmd {
+
+using c10::IValue;
+
+namespace {
+
+template <class T>
+std::vector<std::shared_ptr<Future>> createOneFutureEvent(
+    c10::intrusive_ptr<T> event) {
+  auto future = std::make_shared<Future>(at::AnyClassType::get());
+  future->markCompleted(
+      IValue(c10::static_intrusive_pointer_cast<Event>(event)));
+  std::vector<std::shared_ptr<Future>> futures;
+  futures.reserve(1);
+  futures.emplace_back(std::move(future));
+  return futures;
+}
+
+} // namespace
+
+/////////////////////////////////////////////////////////////////////
+//                         DefaultTrigger                          //
+/////////////////////////////////////////////////////////////////////
+
+std::vector<EventSchema> DefaultTrigger::ingressEvents() const {
+  return {EventType::PREPARE_MODULE, EventType::PRE_FORWARD};
+}
+
+std::vector<EventSchema> DefaultTrigger::egressEvents() const {
+  return {EventType::LOCAL_GRAD_READY};
+}
+
+std::vector<std::shared_ptr<Future>> DefaultTrigger::handleEvent(
+    const c10::intrusive_ptr<Event>& event) {
+  switch (event->schema().type_) {
+    case EventType::PREPARE_MODULE:
+      return handlePrepareModule(
+          c10::static_intrusive_pointer_cast<PrepareModuleEvent>(event));
+    case EventType::PRE_FORWARD:
+      return handlePreForward(
+          c10::static_intrusive_pointer_cast<PreForwardEvent>(event));
+    default:
+      TORCH_INTERNAL_ASSERT(
+          false,
+          "DefaultTrigger retrieved an unexcepted event type ",
+          event->schema().type_);
+  }
+}
+
+std::vector<std::shared_ptr<Future>> DefaultTrigger::handlePrepareModule(
+    c10::intrusive_ptr<PrepareModuleEvent> event) {
+  params_ = event->parameters();
+  for (size_t index = 0; index < params_.size(); ++index) {
+    auto& param = params_[index];
+    auto gradAccumulator = torch::autograd::impl::grad_accumulator(param);
+    // Hook to execute after the gradient accumulator has executed.
+    gradAccumulator->add_post_hook(
+        torch::make_unique<torch::autograd::utils::LambdaPostHook>(
+            [this, index](
+                const torch::autograd::variable_list& outputs,
+                const torch::autograd::variable_list& /* unused */) {
+              autogradHook(index);
+              return outputs;
+            }));
+    gradAccumulators_.push_back(std::move(gradAccumulator));
+  }
+  return {};
```

Comment:
Shouldn't this actually return a `PRE_FORWARD` event type that is handled by another handler?

Reply:
hmm, why? `PREPARE_MODULE` is within the module ctor, and at that time, we haven't reached the forward function yet. 

`PRE_FORWARD` is also a type I event which should be injected by explicitly calling `Engine::preForward`.